PMax Assets: Assets form for managing asset group

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

This PR implements the "Assets form" part of 📌 [Assets form and related components](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-misc) in #1787.

The assets form includes:

- Add `AssetGroupCard` for managing the an asset group
- Validate asset group and show error messages after clicking on the "Launch paid campaign" button
- After clicking on "Launch paid campaign", scroll to the error asset field which is nearest the top of page
- Adjust `CampaignAssetsForm` to validate the campaign values only

Other required changes:

- Add `extendAdapter` prop to `AdaptiveForm` for customizing the adapter of the form context
- Add `className` prop to `AssetField` and also make `numOfIssues` and `subheading` props optional

### Screenshots:

https://user-images.githubusercontent.com/17420811/213707451-e25eea85-b154-4b17-93b3-8dcd682f2458.mp4

### Detailed test instructions:

💡 Please note that this PR doesn't include the API communication of creating/updating the asset group.

1. Go to the campaign creation page.
2. Load an asset group via the final URL.
3. Test the form UI of managing the asset group.
4. Test the form validation and error messages.

### Changelog entry

> Add - Asset group management form for the campaign creation and editing pages.

eason9487 committed:
Add `extendAdapter` prop to AdaptiveForm for customizing the adapter of the form context

eason9487 committed:
Add the form validator for the asset group values

eason9487 committed:
Add the specs of the asset fields

eason9487 committed:
Add `className` prop to AssetField and also make `numOfIssues` prop optional

eason9487 committed:
Implement AssetGroupCard for managing the an asset group

eason9487 committed:
Adjust CampaignAssetsForm to validate the asset group and validate the campaign only

eason9487 committed:
Update asset group values to CampaignAssetsForm

eason9487 committed:
Replace the demos in AssetGroupSection with AssetGroupCard

puntope commented:
💅  Behaviour looking good. However, I'm thinking if it would be better to show the issues in the very beginning instead of when the user performs a change or clicks on Launch Campaign. I guess this is more a UX question (I didn't check if there is any discussion about in P2)

puntope commented on js/src/components/paid-ads/ads-campaign.js:
```diff
@@ -25,8 +25,8 @@ import FaqsSection from './faqs-section';
- * Please note that this component relies on an AdaptiveForm's context, so it expects
- * a context provider component (`AdaptiveForm`) to existing in its parents.
+ * Please note that this component relies on an CampaignAssetsForm's context and custom adapter,
+ * so it expects a `CampaignAssetsForm` to existing in its parents.
```
❓  I can see `<AdsCampaign>` being rendered in `<AdsStepper>` which is rendered by `<SetupAdsForm>` using just `<AdaptiveForm>`. So seems like it's not true that  `<CampaignAssetsForm>`  is a required wrapping component for  `<AdsCampaign>`.

eason9487 replied:
Thanks for catching this. The form component in `SetupAdsForm` indeed needs to be changed to `CampaignAssetsForm` as well, otherwise the "Continue" button in the ads setup flow will always be disabled. Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/1863/commits/716526c20f7b332baf1aafadd249c87d2ee8500f.

puntope commented on js/src/components/paid-ads/asset-group/asset-group-card.js:
```diff
@@ -0,0 +1,236 @@
+/**
+ * Renders the Card UI for managing the an asset group.
+ *
+ * Please note that this component relies on an CampaignAssetsForm's context and custom adapter,
+ * so it expects a `CampaignAssetsForm` to existing in its parents.
```
Since this is only rendered in `<AssetGroupSection>` I would remove this duplicated docs. I take as granted `<AssetGroupCard>` is not going to be used outside  `<AssetGroupSection>`.

puntope replied:
Seems like this is only used in `<AssetGroup>`. So IMO better only in `<AssetGroup>` to avoid duplication.

eason9487 replied:
Added this code comment is not based on the composition of components, but on whether a component uses `useAdaptiveFormContext` and the form adapter extended by `CampaignAssetsForm`.

It's independent and unrelated between components, so it's not a duplication but an indication of the component's dependency on the context provider.

puntope commented on js/src/components/paid-ads/asset-group/asset-group-section.js:
```diff
@@ -2,25 +2,27 @@
 /**
  * Renders the form content for managing an asset group of a campaign with Section UI.
+ *
+ * Please note that this component relies on an CampaignAssetsForm's context and custom adapter,
```
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1863/files#r1087449867

Seems like this is only used in `<AssetGroup>` so I guess we could remove the comment duplication.

eason9487 replied:
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1863#discussion_r1090437704.

puntope commented on js/src/components/paid-ads/constants.js:
```diff
@@ -0,0 +1,282 @@
+	},
+];
+
+{
```
❓  Why we need to wrap this with `{}`

eason9487 replied:
This scoped block is to signal that dynamic initialization is handled here, and that the functions inside are for initialization purposes only and are not expected to be exposed outside this module. I added this as the code comment in https://github.com/woocommerce/google-listings-and-ads/pull/1863/commits/de0abb4063db89f7db3caa396cdee86cb057c711.

puntope commented on js/src/components/paid-ads/constants.js:
```diff
@@ -0,0 +1,282 @@
+	} );
+}
+
+const ASSET_DISPLAY_URL_PATH_SPECS = [
+	{
+		maxCharacterCount: 15,
+		capitalizedName: _x(
+			'The first display URL path',
+			'Capitalized asset field name as the start of an error message',
+			'google-listings-and-ads'
+		),
+	},
+	{
+		maxCharacterCount: 15,
+		capitalizedName: _x(
+			'The second display URL path',
+			'Capitalized asset field name as the start of an error message',
+			'google-listings-and-ads'
+		),
+	},
+];
```
💅 WDYT about moving this one at the top to avoid the dynamic logic and the private functions `getImageHelpContent` and `getSubheading` to be mixed between the constants.

eason9487 replied:
Moved by https://github.com/woocommerce/google-listings-and-ads/pull/1863/commits/03dd7667ef2c109b7490ba83ca15b482f1afc49a.

puntope commented on js/src/components/paid-ads/asset-group/asset-group.js:
```diff
@@ -34,9 +34,17 @@ export default function AssetGroup( { campaign } ) {
+	const { isValidCampaign, isSubmitting, isSubmitted, submitter } = adapter;
 	const currentAction = submitter?.dataset.action;
 
+	const handleLaunchClick = ( ...args ) => {
+		if ( isValidForm ) {
+			handleSubmit( ...args );
+		} else {
+			adapter.showValidation();
+		}
+	};
```
❓  Maybe to discuss with UX if it would be convenient to show the validation after selecting URL and populate assets the first time?

eason9487 replied:
@puntope, thanks for the suggestion.

> Maybe to discuss with UX if it would be convenient to show the validation after selecting URL and populate assets the first time?

> However, I'm thinking if it would be better to show the issues in the very beginning instead of when the user performs a change or clicks on Launch Campaign. I guess this is more a UX question (I didn't check if there is any discussion about in P2) - Quoted from https://github.com/woocommerce/google-listings-and-ads/pull/1863#issuecomment-1403491807

The flow of showing error messages can be found in this design walkthrough video: pcTzPl-1eS-p2#design-walkthrough

https://user-images.githubusercontent.com/17420811/215654469-4357d7bc-b6ef-4f2a-ae63-3b87773cd28e.mp4

The common ways of handling form validation errors are:

1. Errors are displayed after the user leaves an input field.
2. Errors are displayed after the user submits the form.
3. Start displaying errors while the user is entering.

While it's not best practice to display errors after clicking the submit button, displaying errors before the user has entered data or has touched any input fields is considered a bad design (aka premature validation) in many UX recommendations. The reasons for avoiding overly aggressive premature validation are:

- Users may feel frustrated and unfairly reprimanded when a site claims they’ve made mistakes before they’ve had a chance to enter a valid input.
- Users may be distracted from their typing.
- The first glance of the form for users is a bunch of errors that may be very scary.

Therefore, I think it's probably not necessary to discuss this approach with the designer. But if you think this is still the better way to go, I will still discuss it with the designer.

puntope commented on js/src/components/paid-ads/asset-group/asset-group-card.js:
```diff
@@ -0,0 +1,236 @@
+				/>
+			</AssetField>
+			<AssetField
+				ref={ refFirstErrorField.bind( 'display_url_path' ) }
```
💅  WDYT of using a constant instead of hardcoding the string?

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1863/commits/82af988567984bb248bba1d096000f96c30b6971.

puntope commented on js/src/components/paid-ads/asset-group/asset-group-card.scss:
```diff
@@ -0,0 +1,74 @@
+
+.gla-asset-field-call-to-action {
+	.components-base-control {
+		width: 240px;
```
💅  Maybe we can create a new dimension inside `js/src/css/abstracts/_variables.scss` or even use `$gla-width-small` if it's not very problematic to get that 10px extra.

eason9487 replied:
The `$gla-width-small` is a generic name that does not semantically fit the purpose here. Currently, there is only this component that needs this width value for aligning the `<input>` and `<select>`. I tend to move it to **_variables.scss** when there is a more wide reuse.

puntope commented on js/src/components/paid-ads/campaign-assets-form.js:
```diff
@@ -1,28 +1,83 @@
+ * @typedef {import('.~/components/types.js').AssetGroupFormValues} AssetGroupFormValues
  */
 
+const emptyAssetGroup = {
```
💅  Maybe we can rename this in uppercase `EMPTY_ASSET_GROUP`

eason9487 replied:
Unless variables will be exposed outside a module, it does not seem to be a clear convention for variables used only inside a module. Searching this codebase globally with `^const [a-z]\w* = [\[\{]` shows plenty of non-all-uppercase naming.

puntope commented on js/src/components/paid-ads/constants.js:
```diff
@@ -0,0 +1,282 @@
+/**
```
💅  I would name this `specs.js` or `specs-config.js` to be less generic

eason9487 replied:
Renamed to `asset-specs.js` by https://github.com/woocommerce/google-listings-and-ads/pull/1863/commits/856763cd1103f707523f1399a526b7da7baf5ce6.

puntope reviewed (approved):
✅  LGTM

I tested the behaviour as well as the code quality.

Left some non blocking comments. Approved in advance

eason9487 committed:
Fix the campaign form validation for the ads setup flow.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1863#discussion_r1087449292

eason9487 committed:
Move the declaration of `ASSET_DISPLAY_URL_PATH_SPECS` to the top of the file.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1863#discussion_r1087454698

eason9487 committed:
Add a code comment on why there is a scoped block in the js/src/components/paid-ads/constants.js file.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1863#discussion_r1087453951

eason9487 committed:
Rename the constants.js file in js/src/components/paid-ads to asset-specs.js

Address https://github.com/woocommerce/google-listings-and-ads/pull/1863#discussion_r1087576526

eason9487 committed:
Create a constant for the data field 'display_url_path'.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1863#discussion_r1087531719
